Init project

popeskul opened the pull request:
Title: Add health-check and golangci-lint

Description:
This PR adds a health-check endpoint to the server and also integrates golangci-lint for better code quality. 

Changes Made:
- Added a new handler to handle health-check endpoint.
- Configured golangci-lint for the project.
- Fixed some linting errors in the existing code.
- Updated README.md to include information on how to run the health-check endpoint.

Related Issues:
None.

Checklist:
- [x] Added health-check endpoint
- [x] Integrated golangci-lint
- [x] Fixed linting errors
- [x] Updated README.md
- [x] Tested the changes locally
- [x] Created this pull request

Additional Notes:
None.

http://localhost:8080/swagger/index.html

popeskul committed:
init-project-structure: init

nkolosov commented on internal/handlers/v1/health.go:
```diff
@@ -0,0 +1,19 @@
+// @Produce json
+// @Success 200 {string} string
+// @Router /health-check [get]
+func (h *Handlers) HealthCheck(c *gin.Context) {
```
health-check обычно находится на отдельном порту и в отдельном хендлере, т.е. не так: GET :8080//api/v1/health-check, а так: GET :8081/health

nkolosov commented on internal/handlers/v1/user.go:
```diff
@@ -0,0 +1,80 @@
+
+	customer, err := h.service.UserService.FindUserByEmail(c, email)
+	if err != nil {
+		newErrorResponse(c, http.StatusInternalServerError, err.Error())
```
Если пользователь не найден, это не 500-ая ошибка должна быть, а 404

nkolosov commented on internal/handlers/v1/user.go:
```diff
@@ -0,0 +1,80 @@
+
+	createUser, err := h.service.UserService.CreateUser(c, userDTO)
+	if err != nil {
+		newErrorResponse(c, http.StatusInternalServerError, err.Error())
```
Не информативные ошибки.

Например, я отправил пустой JSON на POST
Получил

POST http://localhost:8080/api/v1/users

HTTP/1.1 500 Internal Server Error
Content-Type: application/json; charset=utf-8
Date: Sun, 07 May 2023 07:48:32 GMT
Content-Length: 26

{
  "message": "unknown type"
}

И только по коду смог понять что не так

nkolosov commented on Makefile:
```diff
@@ -0,0 +1,25 @@
+	swag init -g cmd/main.go
+
+test:
+	go test -v ./...
```
Тут бы еще добавить команды для сбора зависимостей и непосредственно компиляции (т.к. собирать docker-образ при локальной разработке - неудобно

nkolosov commented on internal/repository/user.go:
```diff
@@ -0,0 +1,34 @@
+)
+
+type UserStore struct {
+	store memory.UserDB
```
Интерфейс UserDB должен быть рядом с репозиторием, т.е сейчас если нужно будет заменить memory на DB, то придется менять и репозиторий и сам сторадж добавлять, а по идее нужно будет только добавить новый сторадж и передать его в NewUserRepository

popeskul committed:
feat: add support for process.env.PORT environment variable to be able to run app on a configurable port

fix: change port variable case from lowercase port to uppercase PORT to improve semantics

refactor: remove health check endpoint from handlers and add a separate health check server

refactor: move conversion functions from user repository to domain package

refactor: change user repository to use interface instead of concrete implementation

refactor: change user service to use domain conversion functions instead of repository conversion functions

refactor: change user service to use domain package instead of repository package

refactor: change error message in CreateUser handler to be more descriptive

refactor: change error message in GetUserByFilters handler to be more descriptive

refactor: add String() method to User and UserDTO domain objects for better logging

refactor: add Validate() method to UserDTO domain object for better validation

refactor: add build and run targets to Makefile

refactor: add deps target to

feat(whip-round): add whip-round script to the project
